rs/replicated_state/src/canister_state/system_state.rs

```diff
@@ -268,6 +271,219 @@ impl CanisterHistory {
     }
 }
 
+/// `TaskQueue` represents the implementation of a queue structure for tasks.
+#[derive(Clone, Eq, PartialEq, Debug, Default)]
+pub struct TaskQueue {
+    /// Queue of tasks.
+    queue: VecDeque<ExecutionTask>,
+
+    /// Status of low_on_wasm_memory hook execution.
+    on_low_wasm_memory_hook_status: OnLowWasmMemoryHookStatus,
+}
+
+impl TaskQueue {
+    pub fn from_checkpoint(
+        queue: VecDeque<ExecutionTask>,
+        on_low_wasm_memory_hook_status: OnLowWasmMemoryHookStatus,
+        canister_id: &CanisterId,
+    ) -> Self {
+        let queue = TaskQueue {
+            queue,
+            on_low_wasm_memory_hook_status,
+        };
+
+        // Because paused tasks are not allowed in checkpoint rounds when
+        // checking dts invariants that is equivalent to disabling dts.
+        queue.check_dts_invariants(
+            FlagStatus::Disabled,
+            ExecutionRoundType::CheckpointRound,
+            canister_id,
+        );
+
+        queue
+    }
+
+    pub fn front(&self) -> Option<&ExecutionTask> {
+        self.queue.front()
+    }
+
+    pub fn pop_front(&mut self) -> Option<ExecutionTask> {
+        self.queue.pop_front()
+    }
+
+    pub fn enqueue(&mut self, task: ExecutionTask) {
+        match task {
+            ExecutionTask::OnLowWasmMemory(hook_condition_check_result) => {
+                if let Some(is_hook_condition_satisfied) = hook_condition_check_result {
+                    self.on_low_wasm_memory_hook_status
+                        .update(is_hook_condition_satisfied);
+                }
+            }
+            ExecutionTask::Heartbeat
+            | ExecutionTask::GlobalTimer
+            | ExecutionTask::AbortedInstallCode { .. }
+            | ExecutionTask::PausedExecution { .. }
+            | ExecutionTask::PausedInstallCode(_)
+            | ExecutionTask::AbortedExecution { .. } => self.queue.push_front(task),
+        };
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.queue.is_empty()
+    }
+
+    pub fn get_queue(&self) -> &VecDeque<ExecutionTask> {
+        &self.queue
+    }
```

Comment:
Is this used anywhere?

Reply:
Good catch, thanks.

Reply:
A mistake on my side, this function is used when we are creating `CanisterStateBits`, so I just have reverted the deletion of it.

Reply:
Ok, but then once we do the refactoring of how this is encoded into `CanisterStateBits` it will be a bit more clear, since we won't need to get the queue and status separately but will happen as one unit.

Reply:
Yes, I agree with that.